base/strings/basic.jl

```diff
@@ -567,14 +567,19 @@ isascii(c::AbstractChar) = UInt32(c) < 0x80
 ## string map, filter, has ##
 
 function map(f, s::AbstractString)
-    out = IOBuffer(sizehint=sizeof(s))
+    out = StringVector(max(4, sizeof(s)÷sizeof(codeunit(s))))
+    index = UInt(1)
     for c in s
         c′ = f(c)
-        isa(c′, AbstractChar) || throw(ArgumentError(
-            "map(f, s::AbstractString) requires f to return AbstractChar; try map(f, collect(s)) or a comprehension instead"))
-        write(out, c′::AbstractChar)
+        isa(c′, Char) || throw(ArgumentError(
```

Comment:
Can this really be changed for `AbstractString` input?

Reply:
Perhaps not - the existing implementation always returns a `String`, even though it takes an AbstractString. So the function really expects a Char, since AFAIK, you have to add Chars to Strings.

Perhaps it is better to create a more generic method for AbstractString, and then specialize this function for String?

Reply:
> you have to add Chars to Strings

The way the string gets created is via writing `AbstractChar`s to an `io` so as long as the `AbstractChar` supports that. I am very uncertain myself of what the requirements are for `AbstractString` / `AbstractChar` so I usually just leave the generic method in place and add an optimized one for `String`.

Reply:
You are right - I have restored the original function for `AbstractString`, and added the new one for `String`. Also amended the tests to test both implementations.